keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -195,24 +248,42 @@ Some pros of using JSON:
 * Easily parsable and transformable
 * Existing tools for ad-hoc analysis (jq)
 
-Proposed default fields:
-* `ts` - timestamp as Unix time (required, `float64`)
-* `v` - verbosity (required, `int`, default 4)
+We would like to discuss some details of format definition that will guide implementation. Definition will define:
+* Reserved JSON keys that will have special meaning
+* Serialization strategy for different types
+
+#### Reserved JSON keys
+
+We would like to reserve some JSON keys to preserve same meaning for all logs within kubernetes.
+Those keys will have special meaning and should not be used as arguments.
+We would like to keep this set of keys as minimal to reduce possible changes, before we introduce a way to version the log schema.
+
+For simplicity we will use flat key structure without namespacing.
+We will update Kubernetes logging guide to discourage usage of those keys as log arguments.
+We will decide if additional static analysis or runtime validation is needed if such problems arise.
+
+Special JSON keys:
+* `ts` - timestamp as Unix time (required, `float`)
+* `v` - verbosity (required, `int`, default 0)
 * `err` - error string (optional, `string`)
 * `msg` - message (required, `string`)
 
-Default Serialization:
-* primitive types (`int`, `float`) - Compatible with [golang json package](https://golang.org/pkg/encoding/json/#Marshal)
-* namespaced Kubernetes objects (`Pod`, `Deployment`) - JSON object consisting of `name` and `namespace`. Example `{"name": "kubedns", "namespace": "kube-system"}`
-* not namespaced Kubernetes objects (`Node`) - JSON object consisting of `name`. Example `{"name": "cluster1-vm-72x33b8p-34jz"}`
-* arbitrary structure (`http.Request`) - JSON objects consisting of public non-pointer fields.
+#### Serialization strategy
+
+We would like to describe the serialization as it will directly impact results of most logs as it will work as default. What do we want from serialization, it should:
+* Have predictable results and performance
+* Not diverge too much from `text` format (e.g. change formatting, add more information)
 
-Proposed serialization of Kubernetes objects will not include additional fields beside those proposed available in `text` format.
-Discussion for additional fields needed in representation of Kubernetes objects is out of scope of this KEP.
+Proposed Serialization of types (we will expand list of special cases if needed):
```

Comment:
Document string as %q ?

Reply:
We also should define whether we intend to respect `stringer` (method `.String() string`) or not.  Both klog and zap do, so we probably should spec that.

Reply:
Thanks.
Will check how klog and zap do it and update KEP based on that.